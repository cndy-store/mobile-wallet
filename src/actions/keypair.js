import { getItem, setItem, removeItem } from '../lib/keypairStorage';
import { StellarSdk } from '../lib/stellar';
const { Keypair, StrKey } = StellarSdk;

const STORAGE_KEY = 'STELLAR_SECRET';

export const KEYPAIR_STARTS_PROCESSING = 'KEYPAIR_STARTS_PROCESSING';
export const KEYPAIR_ERROR = 'KEYPAIR_ERROR';
export const KEYPAIR_LOAD = 'KEYPAIR_LOAD';
export const KEYPAIR_SAVE = 'KEYPAIR_SAVE';
export const KEYPAIR_REMOVE = 'KEYPAIR_REMOVE';

const chomp = str => str.replace(/\n|\r| /g, '');
const isValidSecret = secret => StrKey.isValidEd25519SecretSeed(secret);

export const keypairProcessing = bool => ({
  type: KEYPAIR_STARTS_PROCESSING,
  isProcessing: bool
});

export const keypairErrored = ({ error }) => ({
  type: KEYPAIR_ERROR,
  error
});

export const keypairLoadSucceded = ({ secret }) => {
  const keypair = secret ? Keypair.fromSecret(secret) : null;

  return {
    type: KEYPAIR_LOAD,
    keypair: keypair,
    error: null
  };
};

export const keypairSaveSucceded = ({ secret }) => {
  const keypair = Keypair.fromSecret(secret);

  return {
    type: KEYPAIR_SAVE,
    keypair: keypair,
    error: null
  };
};

export const keypairRemoveSucceded = () => ({
  type: KEYPAIR_REMOVE,
  keypair: null,
  error: null
});

export const loadKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));
    try {
      const secret = await getItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairLoadSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const saveKeypair = secret => {
  secret = chomp(secret);

  return async dispatch => {
    dispatch(keypairProcessing(true));

    if (!isValidSecret(secret)) {
      const error = new Error('Invalid Secret');
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }

    try {
      await setItem(STORAGE_KEY, secret);
      dispatch(keypairProcessing(false));
      dispatch(keypairSaveSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const removeKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));

    try {
      await removeItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairRemoveSucceded());
      return Promise.resolve();
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};
