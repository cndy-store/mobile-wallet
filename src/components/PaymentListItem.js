import React from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  Text,
  Icon,
  H1,
  Card,
  CardItem,
  Container,
  Header,
  Left,
  ListItem,
  Right,
  Body,
  Title,
  Content,
  Tab,
  Tabs,
  View
} from 'native-base';
import { styles as s } from 'react-native-style-tachyons';
import ShortenedPublicKey from './ShortenedPublicKey';
import { shortFormat } from '../lib/formatter';

const captions = {
  credit: 'From',
  debit: 'To'
};

const amountPrefixes = {
  credit: '+',
  debit: '-'
};

class PaymentListItem extends React.Component {
  static propTypes = {
    type: PropTypes.oneOf(['credit', 'debit']),
    publicKey: PropTypes.string.isRequired,
    amount: PropTypes.string.isRequired,
    createdAt: PropTypes.string.isRequired
  };

  render() {
    const prefix = amountPrefixes[this.props.type];

    return (
      <ListItem>
        <Left>
          <View>
            <H1>
              {prefix}
              {shortFormat(this.props.amount)}
            </H1>
            <Text note>{this.props.createdAt}</Text>
          </View>
        </Left>
        <Right />
      </ListItem>
    );
  }
}

export default PaymentListItem;
