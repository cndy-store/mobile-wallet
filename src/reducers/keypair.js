import {
  KEYPAIR_STARTS_PROCESSING,
  KEYPAIR_ERROR,
  KEYPAIR_LOAD,
  KEYPAIR_SAVE,
  KEYPAIR_REMOVE
} from '../actions/keypair';

const defaultState = {
  isProcessing: false,
  error: null,
  keypair: null
};

export default function keypair(state = defaultState, action) {
  switch (action.type) {
    case KEYPAIR_STARTS_PROCESSING:
      return { ...state, isProcessing: action.isProcessing };
    case KEYPAIR_ERROR:
      return {
        ...state,
        error: action.error,
        isProcessing: action.isProcessing
      };
    case KEYPAIR_LOAD:
    case KEYPAIR_SAVE:
    case KEYPAIR_REMOVE:
      return {
        ...state,
        keypair: action.keypair,
        error: action.error,
        isProcessing: action.isProcessing
      };
    default:
      return state;
  }
}
