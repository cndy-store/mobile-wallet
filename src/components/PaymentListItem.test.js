import React from 'react';
import { render } from '../__tests__/renderer';
import PaymentListItem from './PaymentListItem';

import { publicKey } from '../__tests__/fixtures/keypair';

it('renders correctly with type credit', () => {
  const { toJSON } = render(
    <PaymentListItem
      type="credit"
      publicKey={publicKey}
      amount="12.00"
      createdAt="2018-05-03T12:00:00Z"
    />
  );
  expect(toJSON()).toMatchSnapshot();
});

it('renders correctly with type debit', () => {
  const { toJSON } = render(
    <PaymentListItem
      type="debit"
      publicKey={publicKey}
      amount="12.00"
      createdAt="2018-05-03T12:00:00Z"
    />
  );
  expect(toJSON()).toMatchSnapshot();
});
