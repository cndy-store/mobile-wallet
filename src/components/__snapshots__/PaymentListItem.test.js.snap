// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with type credit 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomWidth": 0.5,
        "borderColor": "#c9c9c9",
        "flexDirection": "row",
        "marginLeft": 16,
        "paddingBottom": 13,
        "paddingRight": 16,
        "paddingTop": 13,
      }
    }
    virtual={undefined}
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "alignSelf": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
      virtual={undefined}
    >
      <View
        style={Object {}}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontSize": 27,
              "lineHeight": 32,
            }
          }
          virtual={undefined}
        >
          +
          12.00
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          note={true}
          style={
            Object {
              "color": "#a7a7a7",
              "fontFamily": "System",
              "fontSize": 14,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          2018-05-03T12:00:00Z
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "alignSelf": "center",
          "flex": 0.28,
          "paddingBottom": null,
          "paddingLeft": null,
          "paddingRight": null,
          "paddingTop": null,
        }
      }
      virtual={undefined}
    />
  </View>
</View>
`;

exports[`renders correctly with type debit 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomWidth": 0.5,
        "borderColor": "#c9c9c9",
        "flexDirection": "row",
        "marginLeft": 16,
        "paddingBottom": 13,
        "paddingRight": 16,
        "paddingTop": 13,
      }
    }
    virtual={undefined}
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "alignSelf": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
      virtual={undefined}
    >
      <View
        style={Object {}}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontSize": 27,
              "lineHeight": 32,
            }
          }
          virtual={undefined}
        >
          -
          12.00
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          note={true}
          style={
            Object {
              "color": "#a7a7a7",
              "fontFamily": "System",
              "fontSize": 14,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          2018-05-03T12:00:00Z
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "alignSelf": "center",
          "flex": 0.28,
          "paddingBottom": null,
          "paddingLeft": null,
          "paddingRight": null,
          "paddingTop": null,
        }
      }
      virtual={undefined}
    />
  </View>
</View>
`;
